### ISSUE TITLE
Question: Other devices (Tolino/Pocketbook)

### ISSUE THREAD

**Issue by WS64** [koreader/koreader#1386]
There are at least two more different reader families that already run Coolreader, these are the devices by Tolino (not sure if CR runs on all of them or just Tolino Vision) and Pocketbook.
Is there any chance to run Koreader on those too?
I got a Pocket book Lux 2 yesterday, and although Coolreader is not bad on this device (better actually than on Kobo devices) I would refer Koreader...

**Comment by WS64** [comment]
Is this correct for a file koreader.app which lies in the same folder as I have put the koreader (Kobo version) folder?
#!/bin/sh
koreader/koreader.sh &
(does not work, "Fehler beim Programmstart")
![image1](https://cloud.githubusercontent.com/assets/7792441/5679457/22f7544a-9800-11e4-809c-19dac7d2a46b.jpg)

**Comment by hwhw** [comment]
Hm, OK. That is how I would have done it. Be sure that this file (koreader.app) does have Unix linefeeds, though. Windows ones will break the shebang (first line). So use Notepad++ or a similar sophisticated editor. Probably, you also need to "exec koreader/koreader.sh" (so no backgrounding via "&"), as the examples do. Maybe, relative paths won't work, depending on the current working directory when the "app" file is executed. You can try something like this then:
```sh
#!/bin/sh
cd $(dirname $0)/koreader
exec ./koreader.sh
```

**Comment by WS64** [comment]
I am aware of the linefeeds and sure i have handled them correctly, but that is about all I know of unix ;)
I tried yours, same error, both for the Kindle and for the Kobo version.

**Comment by hwhw** [comment]
To bad... But looking at koreader.sh, I'm not surprised (sorry, only looking at it now). It just can't work as-is, at it is quite device specific.

If you want to give it one more try, you can do this:
```sh
#!/bin/sh
cd $(dirname $0)/koreader
exec >pbl2.log 2>pbl2-err.log
echo mounts
cat /proc/mounts
./luajit reader.lua -d ./
```
(outputting the filesystem mounts is for testing and info purposes, to get an idea on the environment we run into)

You probably need to try with both the kindle and the kobo files extracted to the koreader subfolder (arm soft float or arm hard float build, which are typically not able to run on the opposite system type).

If we're unlucky, we don't even have necessary rights to create and write to the log files (that exec line will redirect stdout and stderr of the .app script and all subprocesses it launches).

**Comment by WS64** [comment]
Again same msg, and no log files created...
just to make sure i can do anything here at all I copied the sudoku.app and put it there with a new name. no problem with that one.
can it be that this new koreader subfolder is not executable or something like this?

**Comment by hwhw** [comment]
It is possible that the whole filesystem mount is "noexec", i.e. executables on it can't be executed. That said, it doesn't explain why there are no log files. That, however, could be explained by sensible design of the system: The .app files, when executed, do not necessarily run under a user account that is allowed to write to files or create new ones. (Just a short note, if that isn't obvious: The log files would be in the koreader subfolder.)

Can you try a really short one:
```sh
#!/bin/sh
touch test.txt
```

And - if that doesn't create a 0-size file "test.txt", maybe
```sh
#!/bin/sh
cd $(dirname $0)
touch test.txt
```

**Comment by WS64** [comment]
Hmmm, I guess I still had something wrong, probably the linefeeds....
I started all over again, and finally got some results.

Result of

#!/bin/sh
cd $(dirname $0)/koreader
exec >pbl2.log 2>pbl2-err.log
echo mounts
cat /proc/mounts
./luajit reader.lua -d ./

is

pbl2.log:
mounts
rootfs / rootfs rw 0 0
/dev/root / ext2 ro,relatime 0 0
none /proc proc rw,relatime 0 0
none /var tmpfs rw,relatime,size=49152k 0 0
none /sys sysfs rw,relatime 0 0
/dev/mmcblk0p2 /boot vfat ro,dirsync,noexec,noatime,fmask=0022,dmask=0022,codepage=cp437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
none /var/dev/pts devpts rw,relatime,mode=622 0 0
tmpfs /var/dev/shm tmpfs rw,relatime 0 0
/dev/mmcblk0p8 /ebrmain ext2 ro,relatime 0 0
/dev/mmcblk0p9 /mnt/secure ext2 rw,relatime,errors=continue 0 0
/dev/loop0 /ebrmain/cramfs cramfs ro,relatime 0 0
/dev/user_int /mnt/ext1 vfat rw,dirsync,nosuid,nodev,noatime,fmask=0000,dmask=0000,allow_utime=0022,codepage=cp437,iocharset=utf8,shortname=mixed,errors=remount-ro 0 0
/dev/user_extp1 /mnt/ext2 vfat rw,dirsync,nosuid,nodev,noatime,fmask=0000,dmask=0000,allow_utime=0022,codepage=cp437,iocharset=utf8,shortname=mixed,errors=remount-ro 0 0

pbl2-err.log:
/mnt/ext1/applications/wsttest.app: line 6: ./luajit: not found

**Comment by hwhw** [comment]
Ah, great, log files! Hm, isn't there a "luajit" file in the koreader/ directory? Though the error might be misleading, and the "not found" is actually a "not executable". However, the user_int and user_extp1 (sd card? partition 1?) mounts do not have a "noexec" flag set and fmask won't mask out the executable bit, which is also good. So - did you try to switch between kobo/koreader builds yet?

Looks fine otherwise! (You can remove that "cat /proc/mounts" line to keep things short, now that we have established what the situation is)

When it eventually runs (and I do not have doubts that it will in the end!), it will probably stumble upon device recognition. But let's wait and see.

**Comment by WS64** [comment]
The above was Kobo.

Kindle:

pbl2.log:
mounts
rootfs / rootfs rw 0 0
/dev/root / ext2 ro,relatime 0 0
none /proc proc rw,relatime 0 0
none /var tmpfs rw,relatime,size=49152k 0 0
none /sys sysfs rw,relatime 0 0
/dev/mmcblk0p2 /boot vfat ro,dirsync,noexec,noatime,fmask=0022,dmask=0022,codepage=cp437,iocharset=iso8859-1,shortname=mixed,errors=remount-ro 0 0
none /var/dev/pts devpts rw,relatime,mode=622 0 0
tmpfs /var/dev/shm tmpfs rw,relatime 0 0
/dev/mmcblk0p8 /ebrmain ext2 ro,relatime 0 0
/dev/mmcblk0p9 /mnt/secure ext2 rw,relatime,errors=continue 0 0
/dev/loop0 /ebrmain/cramfs cramfs ro,relatime 0 0
/dev/user_int /mnt/ext1 vfat rw,dirsync,nosuid,nodev,noatime,fmask=0000,dmask=0000,allow_utime=0022,codepage=cp437,iocharset=utf8,shortname=mixed,errors=remount-ro 0 0
/dev/user_extp1 /mnt/ext2 vfat rw,dirsync,nosuid,nodev,noatime,fmask=0000,dmask=0000,allow_utime=0022,codepage=cp437,iocharset=utf8,shortname=mixed,errors=remount-ro 0 0

pbl2-err.log:
./luajit: frontend/device.lua:24: did not find a hardware abstraction for this platform
stack traceback:
	[C]: in function 'error'
	frontend/device.lua:24: in function 'probeDevice'
	frontend/device.lua:27: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

**Comment by WS64** [comment]
I guess it is too risky to just 
return require("device/kindle/device")
instead of the error for testing propose?

**Comment by hwhw** [comment]
It isn't risky, but since the code in question will just try to fetch the Kindle model and act to that, the next error is imminent.
I'll push a general first implementation to the code base in a few minutes. It probably will still leave a lot room for try&error, though. We'll probably need to fiddle a bit before refresh & co work properly.

**Comment by hwhw** [comment]
For a quick try - if you feel like it - you can replace frontend/device.lua by this: https://github.com/hwhw/koreader/raw/master/frontend/device.lua and create a new directory frontend/device/pocketbook and put this into it: https://github.com/hwhw/koreader/raw/master/frontend/device/pocketbook/device.lua
Note that the file for now will just pretend to be something like a Kindle Paperwhite.

**Comment by WS64** [comment]
Leads to:

./luajit: ffi/framebuffer_linux.lua:60: framebuffer model not supported
stack traceback:
	[C]: in function 'error'
	ffi/framebuffer_linux.lua:60: in function 'init'
	ffi/framebuffer_mxcfb.lua:265: in function 'init'
	ffi/framebuffer.lua:59: in function 'new'
	frontend/device/pocketbook/device.lua:19: in function 'init'
	frontend/device.lua:38: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

**Comment by WS64** [comment]
btw, just for the record, PB runs this as an application, and the reader does multitasking.
That means that we probably do not have to care for screensaver and light since this is still done by the PocketBook software. But since I can't start Koreader (yet) I am not 100% sure. For Coolreader it works like this.

**Comment by hwhw** [comment]
So it doesn't identify as a "eink_fb" or "mxc_epdc_fb" device. Too bad. I'll look at their source code (which includes the kernel) and will try to establish what kind of eink driver they use.

**Comment by hwhw** [comment]
Their kernel sources seem incomplete to me - at least I can't find the actual config file.

So can you edit "ffi/framebuffer_linux.lua", line 60, to read
```lua
        error("framebuffer model not supported: "..ffi.string(finfo.id))
```
and provide the new error message generated?

**Comment by WS64** [comment]
Hmmm, it's empty:
./luajit: ffi/framebuffer_linux.lua:60: framebuffer model not supported: 
stack traceback:
	[C]: in function 'error'
	ffi/framebuffer_linux.lua:60: in function 'init'
	ffi/framebuffer_mxcfb.lua:265: in function 'init'
	ffi/framebuffer.lua:59: in function 'new'
	frontend/device/pocketbook/device.lua:19: in function 'init'
	frontend/device.lua:38: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

**Comment by WS64** [comment]
tostring(finfo.id)=cdata<char (&)[16]>: 0x400baa10

**Comment by hwhw** [comment]
Well, we just take a guess and use the "mxc_epdc_fb" option:
Line 41 could be changed to
```lua
    if true then
```
But we're really a bit in the dark here.

**Comment by WS64** [comment]
<pre>cdata<char (&)[16]>: 0x400baa10</pre>

**Comment by WS64** [comment]
I check, one moment

**Comment by WS64** [comment]
./luajit: frontend/device/pocketbook/device.lua:23: attempt to index field 'input' (a nil value)
stack traceback:
	frontend/device/pocketbook/device.lua:23: in function 'init'
	frontend/device.lua:38: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

**Comment by hwhw** [comment]
Oh, and maybe add a line
```sh
cat /proc/fb
```
to the koreader.app file, so it will output a (tiny) bit of info on the framebuffer.

**Comment by WS64** [comment]
at the end?

**Comment by hwhw** [comment]
No, better before the call to luajit. I'll debug the problem in frontend/device/pocketbook/device.lua and will update the file soon. It will then have to be exchanged.

**Comment by WS64** [comment]
Or before luajit?

I'll try this:

#!/bin/sh
cd $(dirname $0)/../system/koreader
exec >pbl2.log 2>crash.log
cat /proc/fb
./luajit reader.lua -d ./

**Comment by hwhw** [comment]
Yep. Also, I've just updated the file frontend/device/pocketbook/device.lua in my tree (see link above).

**Comment by WS64** [comment]
pbl2.log just contains a zero. Or where to check for the framebuffer info? crash.log is the same...
./luajit: frontend/device/pocketbook/device.lua:23: attempt to index field 'input' (a nil value)
stack traceback:
	frontend/device/pocketbook/device.lua:23: in function 'init'
	frontend/device.lua:38: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

Ok, now I get the new device.lua and try again

**Comment by hwhw** [comment]
A literal zero ("0"), I guess? That's OK. It means (again, as we likely already found out before) that the framebuffer driver does not provide any name of itself at all. But it tells that there *is* framebuffer 0. Guess we knew that before, too.

**Comment by WS64** [comment]
./luajit: frontend/device/input.lua:147: error opening input device </dev/input/event2>: 13
stack traceback:
	[C]: in function 'open'
	frontend/device/input.lua:147: in function 'open'
	frontend/device/pocketbook/device.lua:32: in function 'init'
	frontend/device.lua:38: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:86: in main chunk
	[C]: at 0x0000bde5

**Comment by WS64** [comment]
And yes, a "0"

**Comment by hwhw** [comment]
new pocketbook/device.lua file just went up!

**Comment by WS64** [comment]
./luajit: ffi/mupdf.lua:21: /lib/libc.so.6: version `GLIBC_2.7' not found (required by libs/libmupdf.so)
stack traceback:
	[C]: in function 'load'
	ffi/mupdf.lua:21: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/imagewidget.lua:4: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/button.lua:3: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/buttontable.lua:5: in main chunk
	[C]: in function 'require'
	frontend/ui/widget/buttondialog.lua:1: in main chunk
	[C]: in function 'require'
	frontend/apps/filemanager/filemanagerhistory.lua:3: in main chunk
	[C]: in function 'require'
	frontend/apps/filemanager/filemanager.lua:1: in main chunk
	[C]: in function 'require'
	reader.lua:135: in main chunk
	[C]: at 0x0000bde5

**Comment by hwhw** [comment]
OK, that's a showstopper, for now, at least. I'm puzzled since GLIBC_2.7 is, well, a symbol present in glibcs since ancient times (Kindle was one of the most outdated embedded device environments I was aware of...). Maybe they use another libc... In any case, we probably need to compile the whole kaboodle via their SDK (meaning: their toolchain). It will take me quite a while to find the time to set that up, unfortunately. Maybe somebody else has some other quick ideas...

Thank you so much for doing all the testing up to this point!

**Comment by WS64** [comment]
I thank you for the quick help!

**Comment by hwhw** [comment]
OK, they seem to use glibc 2.9 (dating back to 2008). Well then. So our "kindle-legacy" build target should probably work... If anyone with a working build environment would like to give it a quick go - "make TARGET=kindle-legacy"... I'll be doing it in 2-4 days otherwise.

**Comment by WS64** [comment]
Any news here?

**Comment by hwhw** [comment]
No, sorry, no spare time yet. With a bit of luck, on the weekend.

**Comment by WS64** [comment]
The new PB release puts the Koreader folder into the main directory of the device. I am not sure if this is a good idea, shouldn't we put Koreader into the system folder? For Kobo we too put it into the .kobo folder, not the main directory. Reason for this proposal is that the PB book browser has access to the Koreader folder but not to the system folder, so Koreader would stay invisible if we put it inside system. I am not sure which is better, but maybe we should at least think about where the Koreader directory should be.

**Comment by chrox** [comment]
Sounds reasonable. The updated preview release has moved `koreader` directory into `applications` directory.

**Comment by WS64** [comment]
Not good either. This way the Koreader directory shows up in the applications menu of the PB.

**Comment by chrox** [comment]
What I observed is that only scripts with ".app" ending in their names in the applications directory are listed in the application menu. And it's pretty safe to put the koreader directory in the applications directory at least in firmware 5.4.

**Comment by WS64** [comment]
I can see all .app files plus all subfolders, so I will for sure NOT keep the Koreader directory in the applications folder. I already have to switch through 3 pages of applications, so I want to keep these list as short as possible.

**Comment by chrox** [comment]
Seems not a problem with firmware 5.4, luckily. :)

**Comment by WS64** [comment]
I just tried the new version for PB, I also got the horizontal lines now.
But, one thing is a bad idea: Don't just overwrite the extension.cfg file. Remove this from the system folder please. My original extensions.cfg file looked like this:
<pre>
acsm:@ACSM_file:1:AdobeViewer.app:ICON_ACSM
app:@APP_file:4:start.app:ICON_APP
asp:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
bmp:@BMP_image:2:jpegviewer.app:ICON_BMP
cgi:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
chm:@Z_HTML_file:1:fbreader.app,cr3-pb.app:ICON_CHM
djv:@DJVU_file:1:djviewer.app:ICON_DJVU
djvu:@DJVU_file:1:djviewer.app:ICON_DJVU
doc:@DOC_file:1:fbreader.app:ICON_DOC
docm:@DOC_file:1:fbreader.app:ICON_DOC
docx:@DOC_file:1:fbreader.app:ICON_DOC
epub:@EPUB_file:1:AdobeViewer.app,fbreader.app,cr3-pb.app:ICON_EPUB
fb2:@FB2_file:1:fbreader.app,cr3-pb.app:ICON_FB2
hrss:@HTML_file:1:links.app:ICON_HTM
htm:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
html:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTML
jpeg:@JPEG_image:2:jpegviewer.app:ICON_JPG
jpg:@JPEG_image:2:jpegviewer.app:ICON_JPG
jsp:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
mht:@HTML_file:1:cr3-pb.app:ICON_HTM
mobi:@PRC_file:1:fbreader.app:ICON_PRC
mp3:@Music_file:3:play.app:ICON_MP3
pbi:@APP_file:4:sysinstall.app:ICON_APP
pdb:@PRC_file:1:fbreader.app:ICON_PRC
pdf:@PDF_file:1:AdobeViewer.app,pdfviewer.app:ICON_PDF
php:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
pl:@HTML_file:1:fbreader.app,cr3-pb.app:ICON_HTM
png:@PNG_image:2:jpegviewer.app:ICON_PNG
prc:@PRC_file:1:fbreader.app:ICON_PRC
rtf:@RTF_file:1:fbreader.app,cr3-pb.app:ICON_RTF
tcr:@TCR_file:1:fbreader.app:ICON_TCR
tif:@TIFF_image:2:jpegviewer.app:ICON_TIFF
tiff:@TIFF_image:2:jpegviewer.app:ICON_TIFF
txt:@Text_file:1:fbreader.app,cr3-pb.app:ICON_TXT
wlnk:@weblink:5:fbreader.app:ICON_WWW
</pre>
Nearly all of this was gone, I could not use Coolreader at all anymore. Not sure how we should handle this, but not by overwriting all these entries....

**Comment by chrox** [comment]
@WS64 This is fixed in #1399. The system directory won't override local existing extensions.cfg file in OTA process now. But it still will be in release package. As the installation instruction [says](https://github.com/koreader/koreader/wiki/Installation-on-PocketBook-devices) "copy the system directory to PocketBook USB root is optional", it still needs users to manually copy to override their existing file. If I recall clearly, Windows users will get a prompt and ask if the user want to override existing file. If the user choose no nothing will happen.
